src/render.c

```diff
@@ -1,22 +1,296 @@
+            int err = create_color_backup(max_pipe_colors, backup);
+            if(err != 0) {
+                return err;
+            }
+        } else {
+            backup->num_colors = 0;
+        }
+    }
+
+    palette->num_colors = max_pipe_colors;
+    palette->colors = malloc(sizeof(*palette->colors) * palette->num_colors);
+    if(!palette->colors)
+        return ERR_OUT_OF_MEMORY;
+
+    // Set palette, either to the value specified in colors or to an HSL sweep.
+    // Note that calls to init_color or init_extended_color must have the
+    // colour index increased by 1 to avoid writing to colour pair 0.
+    for(int i=0; i < max_pipe_colors; i++) {
+        if(!can_change_color() && !direct){
+            // Just use the colours in the default palette if we can't change
+            // colours. The colour is `i + 1` because we don't want to use
+            // the backgroudn colour for the foreground.
+            palette->colors[i] = set_pair(i, i + 1, COLOR_BLACK);
+        }else{
+            int color;
+            if(colors)
+                color = colors[i];
+            else
+                color = hsl2rgb(((float) i) * 360 / max_pipe_colors, 1, 0.5);
+
+            int pair_index;
+            if(direct) {
+                pair_index = set_color_pair_direct(
+                        i + RESERVED_INDIRECT_PAIRS, color);
+            }else{
+                pair_index = set_color_pair_indirect(
+                        i + RESERVED_INDIRECT_PAIRS, color);
+            }
+            if(pair_index < 0)
+                return pair_index;
+            palette->colors[i] = pair_index;
         }
     }
+    return 0;
+}
+
+/// Free memory associated with color palette.
+void palette_destroy(struct palette *palette) {
+    free(palette->colors);
+}
+
+
+/**
+ * Set the colour pair `pair_index` to the given colour in as version agnostic
+ * a way as possible. If `alloc_pair` is available, then `pair_index` is
+ * ignored; otherwise it is passed as the pair ID to `init_pair`.
+ *
+ * Returns -1 on error, or the pair index
+ */
+int set_pair(int pair_index, int fg, int bg) {
+    int retval;
+    int alloced_index = pair_index;
+#if HAVE_ALLOC_PAIR
+    retval = alloc_pair(fg, bg);
+    alloced_index = retval;
+#elif HAVE_EXTENDED_COLOR
+    retval = init_extended_pair(pair_index, fg, bg);
+#else
+    retval = init_pair(pair_index, fg, bg);
+#endif
+    if(retval == ERR)
+        return retval;
+    return alloced_index;
+}
+
+/**
+ * Set the foreground of a colour pair to `color` (in `0xRRGGBB` form).
+ * The index `color_index` will be used as the pair index if
+ * `alloc_pair` is not supported: otherwise, the index returned can be
+ * unrelated.
+ *
+ * Returns the ID of the new pair, or `ERR_CURSES_ERR` on error.
+ *
+ * Note that the color should be zero-indexed. This function internally
+ * adds 1 to it to avoid overwriting any of the default colours.
+ */
+int set_color_pair_indirect(int color_index, int color) {
```

Comment:
I can see `color` has 24 bits, so `unsigned long` should be completely as `int` is only guaranteed to be at least 16 bits and `unsigned` as your note above states it's zero-indexed.